File: Chapter_2/Chapter_2_in_Python/one.py
class Node:
    data = None
    next = None
    head = None
    tmp = None
    new_node = None

    def insert_elements(n):
        new_node = Node()
        if Node.head == None:
            new_node.data = n
            Node.head = new_node
            Node.tmp = new_node
        else:
            new_node.data = n
            Node.tmp.next = new_node
            Node.tmp = Node.tmp.next

    def print_elements():
        currentNode = Node.head;
        while currentNode != None:
            print(currentNode.data)
            currentNode = currentNode.next

    def remove_duplicate_values():
        seen = dict()
        curr = Node.head
        prev = None
        while curr:
            if curr.data in seen:
                 prev.next = curr.next
                 curr = None
            else:
                seen[curr.data] = 1
                prev = curr
            curr = prev.next



if __name__ == '__main__':
    Node.insert_elements(2)
    Node.insert_elements(4)
    Node.insert_elements(1)
    Node.insert_elements(2)
    Node.print_elements()
    Node.remove_duplicate_values()
    print("After removing duplicates")
    Node.print_elements()

File: Chapter_2/Chapter_2_in_C++/one.cpp
/*  2.1
*   Write code to remove duplicates from an unsorted linked list.
*   FOLLOW UP
*   How would you solve this problem if a temporary buffer is not allowed?
*/
#include<bits/stdc++.h>
using namespace std;
class Node
{
public:
    int data;
    Node *next;
};
Node *head;
Node  *newNode;
Node  *tmp;
void insert_elements(int n)
{
    newNode = new Node;
    if(head == NULL)
    {
        newNode->data = n;
        head = newNode;
        tmp = newNode;
    }
    else
    {

        newNode->data = n;
        tmp->next = newNode;
        tmp = tmp->next;
    }

}
void removeDuplicateValue()
{
    unordered_set<int> seen;
    Node *curr = head;
    Node *prev = NULL;
    while (curr != NULL)
    {
        if (seen.find(curr->data) != seen.end())
        {
           prev->next = curr->next;
           delete (curr);
        }
        else
        {
           seen.insert(curr->data);
           prev = curr;
        }
        curr = prev->next;
    }
}
void print_element()
{
    Node *currentNode = head;
    while(currentNode != NULL)
    {
        cout<<currentNode->data<<" ";
        currentNode = currentNode->next;
    }
    cout<<endl;
}
int main()
{
    insert_elements(5);
    insert_elements(5);
    insert_elements(3);
    insert_elements(2);
    insert_elements(3);
    print_element();
    removeDuplicateValue();
    print_element();
    return 0;
}
